### ISSUE TITLE
Possibly add stake to reparameterization and datatrust suggestion

### ISSUE THREAD

**Issue by rbharath** [computablelabs/goest#152]
At present, creating a new challenge requires placing a stake. Other types of candidates don't require stake. This opens up the following attack

- Submit 1000 reparameterization requests each with malicious changes
- Wait till the end of the voting period and vote in one of these reparameterizations

It's hard for the market to defend against this attack since the market's major stakeholders may not have the stake to vote against all the proposals. If the attacker waits till the very end of the voting period, it may be possible to squeeze through a reparam proposal before stakeholders can respond. A similar attack can be levelled against datatrust registration. (Note a similar attack against listing candidates is harder since the datatrust has to authorize a listing candidate before it can be accepted)

Requiring that reparameterization requests and datatrust registration require stake would make these attacks much harder, since the adversary couldn't spam arbitrary numbers of reparam/datatrust-change candidates. This would make it much more feasible for existing stakeholders to vote down bad requests.

This change would require changing https://github.com/computablelabs/goest/blob/master/contracts/Voting.vy#L129, which currently reads
```
if kind == CHALLENGE: # a challenger must successfully stake a challenge
```
to read
```
if kind != APPLICATION: # a listing candidate doesn't need to stake
```

We would also have to change all tests involving datatrust to have the datatrust owner stake. We'd also have to change all reparameterization tests to add stake. In downstream changes, we'd have to change our private test networks to have the datatrust operator have stake.

**Comment by rbharath** [comment]
This proposed change turns out to basically work, with a minor addition that in `Parameterization.vy` and `Datatrust.vy` the candidate stake has to be unlocked in the `Parameterization.resolveReparam` and `Datarust.resolveRegistration` methods.

**Comment by rbharath** [comment]
In addition, the `Voting.transferStake` method was previously only callable by `Listing.vy` with a gate:
```
assert msg.sender == self.listing_address
```
Since `Parameterization.vy` and `Datatrust.vy` need to call this method as well, this is changed to a privilege check:
```
assert self.hasPrivilege(msg.sender)
```

**Comment by rbharath** [comment]
Fixed by #154

**Comment by rbharath** [comment]
Re-implemented by #165 after rollback.
